Document deleting tool

asaunier opened the pull request:
Related to https://github.com/c2corg/v6_api/issues/465

Delete view:
* [x] Add document id to a special table where ES can figure out what doc should be removed from the index
* [x] when removing an image, update the feed entries that reference it
* [x] when removing an image, delete the files from the images storage (see https://github.com/c2corg/v6_images/issues/35 and https://github.com/c2corg/v6_api/blob/merge-documents/c2corg_api/views/image.py#L116-L137)
* [x] remove geoassociations

Unit tests:
* [x] make sure the existing doc/archives test are correct
* [x] check associations have been cleared
* [x] check cache_versions have been updated?
* [x] check the feed has been cleared
* [x] check comments have been cleared?
* [x] check image files have been removed?
* [ ] check ES index is updated => see https://github.com/c2corg/v6_api/issues/629

Add tests for
* [x] articles
* [x] books
* [x] images
* [x] outings
* [x] xreports

tsauerwein commented on c2corg_api/views/document.py:
```diff
@@ -296,6 +296,35 @@ def _put(
+        document = self._get_document(clazz, id, clazz_locale=clazz_locale)
+        lang = document.locales[0].lang
+
+        associations = doc_associations.get_associations(document, lang, False)
```
You are loading all associations to check if a waypoint is still the main waypoint of a route? This could be improved. You could directly query the routes table:

```
def _is_main_waypoint_of_route(waypoint):
    return DBSession.query(
        exists().
        where(Route.main_waypoint_id == waypoint.document_id)
    ).scalar()
```
Similar to https://github.com/c2corg/v6_api/blob/master/c2corg_api/views/association.py#L161

asaunier replied:
OK thanks, but anyway we need to get the list of associations in order to update their ``cache_versions`` when the document has actually been removed, don't we?

tsauerwein replied:
No, this can be done by calling a stored procedure: https://github.com/c2corg/v6_api/blob/494d48d/c2corg_api/models/cache_version.py#L41

tsauerwein commented on c2corg_api/views/document_associations.py:
```diff
@@ -303,3 +304,23 @@ def get_linked_xreports(document, lang):
+                        'This waypoint cannot be deleted because it is the '
+                        'main waypoint of route {}'.format(route.document_id))
+
+        # check if not the only waypoint of a route
```
See also: https://github.com/c2corg/v6_api/blob/master/c2corg_api/views/association.py#L169

tsauerwein commented on c2corg_api/views/document.py:
```diff
@@ -300,14 +301,18 @@ def _delete(self, clazz):
 
-        # TODO: image => get filenames
+        if document.type == IMAGE_TYPE:
+            filenames = DBSession.query(Image.filename). \
```
I think you want to run this query on `ImageArchive`? Because there is only one entry in the `Image` table for a document id.

tsauerwein commented on c2corg_api/views/document.py:
```diff
@@ -311,40 +311,55 @@ def _delete(
+
+        # Remove associations and update cache of formerly associated docs
+        doc_associations.remove_associations(document)
+        update_cache_version(document)
```
I guess you have to call `update_cache_version` before removing the associations. Otherwise the cache versions of the associations cannot be updated.

tsauerwein commented on c2corg_api/views/document_delete.py:
```diff
@@ -289,6 +290,5 @@ def _remove_associations(document_id):
-    # TODO
-    pass
+def _notify_es_syncer(document_id):
+    DBSession.add(ESDeletedDocument(document_id=document_id))
```
You will also have to send a message to Redis if the database transaction is successful. See the existing function  `notify_es_syncer`.

asaunier commented:
I have rebased the PR to group and rename some commits to make them a bit more readable.
The first one remains a bit big because I have changed my mind, moving most of the code and tests to dedicated files and I had to group the change to keep only the final approach.

A few items are still not addressed in the TODO:
* testing the cache versions => perhaps only the cache version of the deleted doc? Or also the versions of formerly associated docs?
* testing the "removal" of comments (actually only the "associations" in the guidebook's topic table, comments should be removed separately from Discourse).
* testing ES index has actually been updated. I don't know how to test this yet.

asaunier commented:
@tsauerwein Do you mind having a look at this PR? Thanks :)

I'm currently working on the UI side. This will also allow to make real data testing.

tsauerwein commented on c2corg_api/views/document_delete.py:
```diff
@@ -0,0 +1,327 @@
+        _remove_from_feed(document_id)
+
+        if document_type == IMAGE_TYPE:
+            # Remove this image references from the feed
```
"remove the references of this image ..."?

asaunier replied:
Yes, in feed entries you have references to images, see https://github.com/c2corg/v6_api/blob/master/c2corg_api/models/feed.py#L152-L163

If you delete one of these images and don't update the feed entries that reference it, you'll get broken images/links in the remaining feed cards. I have tried to address this problem with https://github.com/c2corg/v6_api/blob/78e8fdc0445213a283a4eedf136a56afa0054369/c2corg_api/views/document_delete.py#L284-L305

asaunier replied:
Oh, perhaps you meant my sentence was incorrect?

tsauerwein replied:
> Oh, perhaps you meant my sentence was incorrect?

Yes. ;)

tsauerwein commented on c2corg_api/views/document_delete.py:
```diff
@@ -0,0 +1,327 @@
+        item.image3_id = None
+        item.more_images = False
+        # FIXME what happens if the deleted image was the only image reference
+        # in the feed item => should the item be removed or replaced?
```
If the change type is `'added_photos'` it would probably make sense to remove the entire feed entry.

tsauerwein commented on c2corg_api/views/document_delete.py:
```diff
@@ -0,0 +1,327 @@
+        # FIXME what happens if the deleted image was the only image reference
+        # in the feed item => should the item be removed or replaced?
+    if len(items):
+        DBSession.flush()
```
Is the flush necessary here?

asaunier replied:
I don't know. I wanted to make sure the changes done in the feed items were actually saved. If you think the flush is useless, I will remove it.

tsauerwein commented on c2corg_api/scripts/es/sync.py:
```diff
@@ -247,6 +264,24 @@ def sync_documents(session, changed_documents, batch_size):
+                '_op_type': 'delete'
+            })
+            n += 1
+        log.info('Removed {} document(s)'.format(n))
```
Maybe move the log statement outside the `with batch:` block. Because the (last) request to ES is only send when the batch is closed. And only then you know if it was successful.

tsauerwein commented on c2corg_api/tests/views/test_document_merge.py:
```diff
@@ -261,6 +255,11 @@ def image_service_mock(url, request):
             }, 200)
             self.assertEqual(call['times'], 1)
+            self.assertIn('filenames=image1.1.jpg', call['request.body'])
+            self.assertIn('filenames=image1.jpg', call['request.body'])
```
Thanks for fixing this test!

tsauerwein commented on c2corg_api/tests/views/test_document_delete.py:
```diff
@@ -0,0 +1,562 @@
+            'changed filename', [UpdateType.FIGURES], [])
+        self.session.flush()
+
+        # TODO test deleting SVG images?
```
Handling this should be in the responsibility of the image service: https://github.com/c2corg/v6_images/blob/master/tests/inside/test_views.py#L27

asaunier replied:
OK. Then I will remove this part if it does not need to be tested here.

tsauerwein commented on c2corg_api/tests/views/test_document_delete.py:
```diff
@@ -0,0 +1,562 @@
+        response = self._delete(self.waypoint1.document_id, 400)
+        errors = response.json.get('errors')
+        self.assertEqual(len(errors), 1)
+        self.assertEqual(errors[0].get('name'), 'Bad Request')
```
There is a convenience method to make testing for errors simpler. For example you could write this test as:

```python
self.assertErrorsContain(response.json, 'Bad Request', 'This waypoint cannot be deleted because it is a main waypoint.')
```

tsauerwein commented on c2corg_api/tests/views/test_document_delete.py:
```diff
@@ -0,0 +1,562 @@
+            AreaAssociation.document_id == document_id).count()
+        self.assertEqual(0, association_count)
+
+        # TODO check cache_versions have been updated
```
For example when deleting `self.waypoint3` you could test that the cache version of the associated route `self.route3` has changed:

```python
        self.check_cache_version(self.route3.document_id, 2)

```

tsauerwein commented:
> check ES index is updated

You could do a test similar to this: https://github.com/c2corg/v6_api/blob/master/c2corg_api/tests/scripts/es/test_sync.py

asaunier commented:
> You could do a test similar to this: https://github.com/c2corg/v6_api/blob/master/c2corg_api/tests/scripts/es/test_sync.py

You mean the full file or adding some test to it?

tsauerwein commented:
> You mean the full file or adding some test to it?

The tests in this file are for the functions in [sync.py](https://github.com/c2corg/v6_api/blob/master/c2corg_api/scripts/es/sync.py). So you could add tests for the functions you've added to this file.

asaunier commented:
Argh!
When I try to delete in a real UI instance an outing with archived versions, I get the following error:
```
sqlalchemy.exc.IntegrityError: (psycopg2.IntegrityError) update or delete on table "documents" violates foreign key constraint "documents_archives_redirects_to_fkey" on table "documents_archives"
DETAIL:  Key (document_id)=(734972) is still referenced from table "documents_archives".
 [SQL: 'DELETE FROM guidebook.documents WHERE guidebook.documents.document_id = %(document_id_1)s'] [parameters: {'document_id_1': 734972}]
```

asaunier commented:
I don't really have time to add tests about ES index syncing when documents are deleted. By the way I don't really know how to simulate deleted docs in https://github.com/c2corg/v6_api/blob/master/c2corg_api/tests/scripts/es/test_sync.py
@tsauerwein would you have an idea?

I have created a dedicated issue for this task https://github.com/c2corg/v6_api/issues/629 and will merge the PR soon if no one objects since @tsauerwein has approved it and since his last comments have been addressed.

asaunier commented:
I have also noticed that the image files are not actually deleted. See https://github.com/c2corg/v6_api/issues/630
